go/adbc/driver/bigquery/connection.go

```diff
@@ -0,0 +1,868 @@
+	case OptionValueAuthTypeJSONCredentialFile, OptionValueAuthTypeJSONCredentialString, OptionValueAuthTypeUserAuthentication:
+		var credentials option.ClientOption
+		if c.authType == OptionValueAuthTypeJSONCredentialFile {
+			credentials = option.WithCredentialsFile(c.credentials)
+		} else if c.authType == OptionValueAuthTypeJSONCredentialString {
+			credentials = option.WithCredentialsJSON([]byte(c.credentials))
+		} else {
+			if c.clientID == "" {
+				return adbc.Error{
+					Code: adbc.StatusInvalidArgument,
+					Msg:  fmt.Sprintf("The `%s` parameter is empty", OptionStringAuthClientID),
+				}
+			}
+			if c.clientSecret == "" {
+				return adbc.Error{
+					Code: adbc.StatusInvalidArgument,
+					Msg:  fmt.Sprintf("The `%s` parameter is empty", OptionStringAuthClientSecret),
+				}
+			}
+			if c.refreshToken == "" {
+				return adbc.Error{
+					Code: adbc.StatusInvalidArgument,
+					Msg:  fmt.Sprintf("The `%s` parameter is empty", OptionStringAuthRefreshToken),
+				}
+			}
+			credentials = option.WithTokenSource(c)
+		}
+
+		client, err := bigquery.NewClient(ctx, c.projectID, credentials)
+		if err != nil {
+			return err
+		}
+
+		err = client.EnableStorageReadClient(ctx, credentials)
+		if err != nil {
+			return err
+		}
+
+		c.client = client
+	default:
+		client, err := bigquery.NewClient(ctx, c.projectID)
+		if err != nil {
+			return err
+		}
+
+		err = client.EnableStorageReadClient(ctx)
+		if err != nil {
+			return err
+		}
+
+		c.client = client
+	}
+	return nil
+}
+
+var (
+	sanitizedInputRegex = regexp.MustCompile("^[a-zA-Z0-9_-]+")
+)
+
+func sanitize(value string) (string, error) {
+	if value == "" {
+		return value, nil
+	} else {
+		if sanitizedInputRegex.MatchString(value) {
+			return value, nil
+		} else {
+			return "", adbc.Error{
+				Code: adbc.StatusInvalidArgument,
+				Msg:  fmt.Sprintf("invalid characters in value `%s`", value),
+			}
+		}
+	}
+}
+
+func buildSchemaField(name string, typeString string) (arrow.Field, error) {
+	index := strings.Index(name, "(")
+	if index == -1 {
+		index = strings.Index(name, "<")
+	} else {
+		lIndex := strings.Index(name, "<")
+		if index < lIndex {
+			index = lIndex
+		}
+	}
+
+	if index == -1 {
+		return buildField(name, typeString, typeString, index)
+	} else {
+		dataType := typeString[:index]
+		return buildField(name, typeString, dataType, index)
+	}
+}
+
+func buildField(name, typeString, dataType string, index int) (arrow.Field, error) {
+	// https://cloud.google.com/bigquery/docs/reference/storage#arrow_schema_details
+	field := arrow.Field{
+		Name: strings.Clone(name),
+	}
+	switch dataType {
+	case "BOOL", "BOOLEAN":
+		field.Type = arrow.FixedWidthTypes.Boolean
+	case "INTEGER", "INT64":
+		field.Type = arrow.PrimitiveTypes.Int64
+	case "FLOAT", "FLOAT64":
+		field.Type = arrow.PrimitiveTypes.Float64
+	case "BYTES":
+		field.Type = arrow.BinaryTypes.Binary
+	case "STRING", "GEOGRAPHY", "JSON":
+		field.Type = arrow.BinaryTypes.String
```

Comment:
The field's metadata should maintain this value as the original type so users don't lose that information.

Honestly, I'd argue that we should always unambiguously add the original type as the metadata